modules/sockets/sockets.cpp

```diff
@@ -15,233 +11,370 @@
-    #ifdef _WIN32 // On windows, check whether the sockets are initialized correctly
-    closesocket(socket);
-    #else
-    // Close the socket (linux/unix styled systems)
-    close(socket);
-    #endif
-    return 0;
+	int length = 0;
+	char *hostname = MF_GetAmxString(amx, params[1], 0, &length);
+
+	cell *error = MF_GetAmxAddr(amx, params[4]);
+	*error = 0;
+
+	if(length == 0)
+	{
+		*error = ERROR_EHOSTUNREACH;
+		return -1;
+	}
+
+	char port_number[6];
+	ke::SafeSprintf(port_number, sizeof(port_number), "%d", params[2]);
+
+	int sockfd = -1, getaddrinfo_status = -1, connect_status = -1, setnonblocking_status = -1;
+	bool connect_inprogress = false;
+	struct addrinfo hints, *server_info, *server;
+
+	memset(&hints, 0, sizeof(hints));
+
+	// Both hostname and port should be numeric to prevent the name resolution service from being called and potentially blocking the call for a long time
+	hints.ai_flags = AI_NUMERICHOST | AI_NUMERICSERV; 
+	hints.ai_family = AF_UNSPEC;
+	hints.ai_socktype = params[3];
+
+	if((getaddrinfo_status = getaddrinfo(hostname, port_number, &hints, &server_info)) != 0)
+	{
+		*error = getaddrinfo_status;
+		return -1;
+	}
+
+	server = server_info;
+
+	do
+	{
+		if((sockfd = socket(server->ai_family, server->ai_socktype, server->ai_protocol)) != -1)
+		{
+			setnonblocking_status = set_nonblocking(sockfd);
+
+			if(setnonblocking_status == 0)
+			{
+				if((connect_status = connect(sockfd, server->ai_addr, server->ai_addrlen)) == -1)
+				{
+					*error = errno;
+
+					if(*error == EINPROGRESS)
+						connect_inprogress = true;
+					else
+						close(sockfd);
+				}
+				else
+				{
+					*error = 0;
+				}
+			}
+			else
+			{
+				if(*error == 0)
+					*error = setnonblocking_status;
+			}
+		}
+		else
+		{
+			if(*error == 0)
+				*error = errno;
+		}
+
+	} while(connect_inprogress == false && connect_status != 0 && (server = server->ai_next) != nullptr);
+
+	freeaddrinfo(server_info);
+
+	if(sockfd == -1 || server == nullptr)
+		return -1;
+
+	return sockfd;
 }
 
-// native socket_change(_socket, _timeout=100000);
-// 1 sec =1000000 usec
-static cell AMX_NATIVE_CALL socket_change(AMX *amx, cell *params)  /* 2 param */
+// native socket_close(_socket);
+static cell AMX_NATIVE_CALL socket_close(AMX *amx, cell *params)
 {
-    int socket = params[1];
-    unsigned int timeout = params[2];
-    //PRINT_CONSOLE("Function: Change | Socket: %i | Timeout: %i\n", socket, timeout);
-    // We need both a timeout structure and a fdset for our filedescriptor
-    fd_set rfds;
-    struct timeval tv;
-    // Fill in ...
-    FD_ZERO(&rfds);
-    FD_SET(socket, &rfds);
-    tv.tv_sec = 0;
-    tv.tv_usec = timeout;
-    // Now we "select", which will show us if new data is waiting in the socket's buffer
-    if (select(socket+1, &rfds, NULL, NULL, &tv) > 0) 
-       return 1; // Ok, new data, return it
-     else 
-       return 0; // No new data, return it
+	return (close(params[1]) == -1) ? -1 : 1;
```

Comment:
Yes, as I said before, having consistency is good. Return values should be standardized across the module, so maybe changing all of them to 0 would be better in order to maintain backwards compatibility with socket_change().